File: 10-heap/01064-Last_Stone_Weight/01064-last_stone_weight.java
// 1064: Last Stone Weight
// https://leetcode.com/problems/last-stone-weight

import java.util.PriorityQueue;


class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        for (int stone : stones) pq.offer(stone);

        while (pq.size() > 1) {
            int y = pq.poll();
            int x = pq.poll();

            if (x != y) {
                pq.offer(y - x);
            }
        }

        return pq.isEmpty() ? 0 : pq.peek();
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] stones = {2,7,4,1,8,1};

        // OUPUT
        var result = o.lastStoneWeight(stones);
        System.out.println(result);
    }
}

File: 10-heap/01064-Last_Stone_Weight/01064-last_stone_weight.py
# 1064: Last Stone Weight
# https://leetcode.com/problems/last-stone-weight


import heapq

class Solution:
    def lastStoneWeight(self, stones: list[int]) -> int:
        stones = [-stone for stone in stones]
        heapq.heapify(stones)

        while len(stones) > 1:
            y = heapq.heappop(stones)
            x = heapq.heappop(stones)
        
            if x != y:
                heapq.heappush(stones, y - x)
        
        return -stones[0] if stones else 0


if __name__ == "__main__":
    o = Solution()

    # INPUT
    stones = [2,7,4,1,8,1]

    # OUTPUT
    result = o.lastStoneWeight(stones)
    print(result)
